README.md
# Long War of the Chosen (LWOTC)

This is a port of [Pavonis Interactive](https://www.pavonisinteractive.com/)'s
Long War 2 (LW2) overhaul mod for XCOM 2 to the War of the Chosen (WOTC)
expansion and update it for WOTC's new features.

## Installing and playing the mod

If you just want to install and play the mod, check out our [wiki page](https://github.com/long-war-2/lwotc/wiki/Installing-Long-War-of-the-Chosen) that explains exactly how to do that.

## Contributing translations

If you would like to contribute to translations for LWOTC, then check out the
[wiki page](https://github.com/long-war-2/lwotc/wiki/Contributing#localization-translating-text-in-the-game)
that explains how it works.

## Building and running the mod

If you want to contribute changes to code or assets, then you will need to
build the mod so that you can test them. Before you can do that, you need to
set some things up:

 1. Make sure you have the WOTC SDK `full_content` branch installed - see the
    [xcom2mods wiki](https://www.reddit.com/r/xcom2mods/wiki/index#wiki_setting_up_tools_for_modding)
    for details on how to do that (plus lots of other useful information)

 2. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo)
    and then clone your fork locally, which you can do via [Git for Windows](https://gitforwindows.org/)
    (a command-line tool), [GitHub Desktop](https://desktop.github.com/), or some other
    git client tool

 3. Once you have cloned the repository, you may need to pull the code for the embedded
    highlander. If the *X2WOTCCommunityHighlander* directory is empty, then use the
    command line from the project's root directory (the one containing this README.md):
    ```
        > git submodule update --init
    ```
    or whatever is the equivalent with the git client tool you are using.

 4. Download the LWOTC media assets (video, graphics and sound) from
    [this Dropbox link](https://www.dropbox.com/s/209rybpdl3khb26/lwotc-content.zip?dl=0)
    and unpack the resulting zip file into this project's *LongWarOfTheChosen* directory.

 5. Set up the following environment variables:
    * `XCOM2SDKPATH` — typically &lt;path to Steam&gt;\steamapps\common\XCOM 2 War Of The Chosen SDK
    * `XCOM2GAMEPATH` — typically &lt;path to Steam&gt;\steamapps\common\XCOM 2\XCom2-WarOfTheChosen
    Don't put these paths in quotes.
	
 6. Open a new command prompt after setting those environment variables and run
    the following from the LWOTC project directory:
    ```
    > build-lwotc.bat -config release
    ```
 
 7. You should also build the Community Highlander, which you can do by opening
    the solution file in X2WOTCCommunityHighlander in Mod Buddy and using that
    to build the project, or you can open the LWOTC project directory in VS Code
    and use the "Terminal > Run Task..." menu option and select "Build CHL
    (final release)" and then "Build DLC2 CHL" once the previous task has finished.

Once the highlander and LWOTC are built, you will be able to select them as local
mods in Alternative Mod Launcher and run Long War of the Chosen.

## Contributing

Contributions are welcome. If you just want to raise issues, please do so [on GitHub](https://github.com/long-war-2/lwotc/issues),
preferably including a save file if possible.

If you wish to contribute to development — and this project will rely heavily on such contributions — then please
look through the issues and if you want tackle one, just leave a comment along the lines of "I'll take this one".
If you find you can't complete the issue in a reasonable time, please add another comment that says you're relinquishing
the issue.

All contributions are welcome, but bug fixes are _extremely_ welcome!

## Acknowledgements

 * Track Two, who has provided a huge amount of advice and insight that saved me lots of time
   and ensured certain bugs got fixed at all.
 * The folks behind X2WOTCCommunityHighlander.
 * All the folks in XCOM 2 modders' Discord who have answered my questions.
 * All the authors of the mods that are integrated into this port:
   - robojumper's Squad Select
   - Detailed Soldier List
 * The Long War 2 team for producing the mod in the first place!


CHANGELOG.md
## Next

## 1.1.0 (2021-06-15)

* Remove compiled script packages when switching between debug and release mode to prevent compiler error (#16)
* Remove compiled script packages when modifying macros (#20)
* Overridden Steam UGC IDs can now be `long` (`int64`) (#22)
* Use error syntax `file(line)` for compiler errors to be compatible with both ModBuddy and VS Code (#26)
* Add a `clean.ps1` script, ModBuddy configuration and VS Code example task to remove all cached build artifacts (#24)
* Remove project file verification. Consider using [Xymanek/X2ProjectGenerator](https://github.com/Xymanek/X2ProjectGenerator) instead (#28)
* Catch macro name clashes through `extra_globals.uci` (#30)
* Add debugging option to profile build times (#35)

## 1.0.0 (2021-05-22)

* Initial release


clean.ps1
Param(
    [string] $modName, # mod folder name
    [string] $srcDirectory, # the path that contains your mod's .XCOM_sln
    [string] $sdkPath, # the path to your SDK installation ending in "XCOM 2 War of the Chosen SDK"
    [string] $gamePath # the path to your XCOM 2 installation ending in "XCOM2-WaroftheChosen"
)

$ErrorActionPreference = "Stop"
Set-StrictMode -Version 3.0

if ($null -eq $modName -or $modName -eq "") {
    throw "`$modName empty???"
}

Write-Host "Deleting all cached build artifacts..."

$files = @(
    "$sdkPath\XComGame\lastBuildDetails.json",
    "$sdkPath\XComGame\Script\*.u",
    "$sdkPath\XComGame\ScriptFinalRelease\*.u",
    "$sdkPath\XComGame\Content\LocalShaderCache-PC-D3D-SM4.upk"
)

$folders = @(
    "$srcDirectory\BuildCache",
    "$sdkPath\Development\Src\*",
    "$sdkPath\XComGame\Mods\*",
    "$gamePath\XComGame\Mods\$modName"
)

$files | ForEach-Object {
    Write-Host "Removing file(s) $($_)"
    Remove-Item -Force $_ -WarningAction SilentlyContinue -ErrorAction SilentlyContinue
}

$folders | ForEach-Object {
    Write-Host "Removing folders(s) $($_)"
    Remove-Item -Recurse -Force $_ -WarningAction SilentlyContinue -ErrorAction SilentlyContinue
}

Write-Host "Cleaned."


.scripts/build.ps1
Param(
    [string] $srcDirectory, # the path that contains your mod's .XCOM_sln
    [string] $sdkPath, # the path to your SDK installation ending in "XCOM 2 War of the Chosen SDK"
    [string] $gamePath, # the path to your XCOM 2 installation ending in "XCOM2-WaroftheChosen"
    [string] $config # build configuration
)

$ScriptDirectory = Split-Path $MyInvocation.MyCommand.Path
$common = Join-Path -Path $ScriptDirectory "X2ModBuildCommon\build_common.ps1"
Write-Host "Sourcing $common"
. ($common)

$builder = [BuildProject]::new("LongWarOfTheChosen", $srcDirectory, $sdkPath, $gamePath)

switch ($config)
{
    "debug" {
        $builder.EnableDebug()
    }
    "default" {
        # Nothing special
    }
    "" { ThrowFailure "Missing build configuration" }
    default { ThrowFailure "Unknown build configuration $config" }
}

$builder.IncludeSrc("$srcDirectory\X2WOTCCommunityHighlander\X2WOTCCommunityHighlander\Src")
$builder.SetContentOptionsJsonFilename("ContentOptions.json")
$builder.InvokeBuild()


.scripts/X2ModBuildCommon/InvokePowershellTask.cs
using System;
using System.IO;
using System.Threading;
using System.Management.Automation;

using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;

public class InvokePowershellTask : Task, ICancelableTask
{
    [Required] public string EntryPs1 { get; set; }
    [Required] public string SolutionRoot { get; set; }
    [Required] public string SdkInstallPath { get; set; }
    [Required] public string GameInstallPath { get; set; }
    [Required] public ITaskItem[] AdditionalArgs { get; set; }

    private PowerShell _ps;

    private ManualResetEventSlim _startingMre = new ManualResetEventSlim(false);

    public override bool Execute()
    {
        bool isSuccess = false;

        try
        {
            _ps = PowerShell.Create();

            _ps
                .AddCommand("Set-ExecutionPolicy")
                .AddArgument("Unrestricted")
                .AddParameter("Scope","CurrentUser");
            
            _ps
                .AddStatement()
                .AddCommand(EntryPs1)
                .AddParameter("srcDirectory", TrimEndingDirectorySeparator(SolutionRoot))
                .AddParameter("sdkPath", TrimEndingDirectorySeparator(SdkInstallPath))
                .AddParameter("gamePath", TrimEndingDirectorySeparator(GameInstallPath));

            foreach (ITaskItem Arg in AdditionalArgs)
            {
                string Val = Arg.GetMetadata("Value");
                if (string.IsNullOrEmpty(Val))
                {
                    _ps.AddParameter(Arg.ItemSpec);
                }
                else
                {
                    _ps.AddParameter(Arg.ItemSpec, Val);
                }
            }

            BindStreamEntryCallback(_ps.Streams.Debug, record => LogOutput(record.ToString()));
            BindStreamEntryCallback(_ps.Streams.Information, record => LogOutput(record.ToString()));
            BindStreamEntryCallback(_ps.Streams.Verbose, record => LogOutput(record.ToString()));
            BindStreamEntryCallback(_ps.Streams.Warning, record => LogOutput(record.ToString())); // TODO: More flashy output?

            BindStreamEntryCallback(_ps.Streams.Error, record =>
            {
                // TODO: Less info than when from console
                // TODO: More flashy output?
                LogOutput(record.ToString());
                Log.LogError(record.ToString());
                isSuccess = false;
            });

            _ps.InvocationStateChanged += (sender, args) => 
            {
                if (args.InvocationStateInfo.State == PSInvocationState.Running)
                {
                    _startingMre.Set();
                }
            };

            isSuccess = true;
            _ps.Invoke();
        }
        catch (System.Exception e)
        {
            Log.LogError(e.Message);
            isSuccess = false;
        }

        return isSuccess;
    }

    public void Cancel()
    {
        // Log.LogMessage(MessageImportance.High, "Got cancel");

        // Do not call Stop() until we know that we've actually started
        // This could be more elaborate, but the time interval between Execute() and Invoke() being called is extremely small

        _startingMre.Wait();
        _ps.Stop();
    }

    private void LogOutput (string output)
    {
        // This is required to keep the empty lines in the output
        if (string.IsNullOrEmpty(output)) output = " ";

        Log.LogMessage(MessageImportance.High, output);
    }

    private static readonly char[] DirectorySeparatorsForTrimming = new char[]
    {
        Path.DirectorySeparatorChar,
        Path.AltDirectorySeparatorChar
    };

    private static string TrimEndingDirectorySeparator(string path)
    {
        return path.TrimEnd(DirectorySeparatorsForTrimming);
    }

    private static void BindStreamEntryCallback<T>(PSDataCollection<T> stream, Action<T> handler)
    {
        stream.DataAdded += (object sender, DataAddedEventArgs e) => handler(stream[e.Index]);
    }
}

.scripts/install_loc.ps1
Param(
    [string] $srcDirectory, # the path that contains your mod's .XCOM_sln
    [string] $sdkPath, # the path to your SDK installation ending in "XCOM 2 War of the Chosen SDK"
    [string] $gamePath # the path to your XCOM 2 installation ending in "XCOM2-WaroftheChosen"
)

Set-StrictMode -Version 3.0

$modNameCanonical = "LongWarOfTheChosen"
$modSrcLocRoot = "$srcDirectory\$modNameCanonical\Localization"
$stagingPath = "$sdkPath\XComGame\Mods\$modNameCanonical\Localization"
$installPath = "$gamePath\XComGame\Mods\$modNameCanonical\Localization"

Write-Host "Copying localization files to staging..."
Robocopy.exe "$modSrcLocRoot" "$stagingPath" *.* "/S" "/NFL" "/NDL" "/NS" "/NC" "/E" "/DCOPY:DA" "/COPY:DAT" "/PURGE" "/MIR" "/NP" "/R:1000000" "/W:30"

Write-Host "Converting to UTF-16 LE..."
Get-ChildItem "$stagingPath" -Recurse -File | 
Foreach-Object {
    $content = Get-Content $_.FullName -Encoding UTF8
    $content | Out-File $_.FullName -Encoding Unicode
}

Write-Host "Copying converted files to the installed mod..."
Robocopy.exe "$stagingPath" "$installPath" *.* "/S" "/NFL" "/NDL" "/NS" "/NC" "/E" "/DCOPY:DA" "/COPY:DAT" "/PURGE" "/MIR" "/NP" "/R:1000000" "/W:30"

Write-Host "Done."
